### PR TITLE
Feature/devise user

### PR DESCRIPTION
**Pull request by abdullah-siddiqui-git**
Implementation of Users module through Devise. Email Confirmation and Invitation are also included.

### CODE REVIEW

**Comment by abdullah-siddiqui-git** [comment]
Review: Cycle-0

**Comment by MidhatAsif** [comment]
https://app.asana.com/0/1132235786988333/1132235786988346/f

File: Gemfile.lock [hunk 1]
```diff
@@ -48,6 +48,7 @@ GEM
     builder (3.2.3)
     byebug (10.0.2)
     cancancan (3.0.1)
+    coderay (1.1.2)
     coffee-rails (4.1.1)
       coffee-script (>= 2.2.0)
```
Comment (MidhatAsif) [inline Gemfile.lock]: dummy comment

File: Gemfile.lock [hunk 2]
```diff
@@ -83,13 +84,19 @@ GEM
       nokogiri (>= 1.5.9)
     mail (2.7.1)
       mini_mime (>= 0.1.1)
+    method_source (0.9.2)
     mini_mime (1.0.2)
```
Comment (MidhatAsif) [inline Gemfile.lock]: dummyyy

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -4,19 +4,21 @@
...
   <%= javascript_include_tag 'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js' %>
+  <!-- END - BOOTSTRAP 4 COMPONENTS -->
 
   <%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>
 
-
   <%= csrf_meta_tags %>
 </head>
 <body>
 <div class="container">
```
Comment (MidhatAsif) [inline app/views/layouts/application.html.erb]: <div> tag in body is to be indented

File: app/controllers/users_custom_controller.rb [hunk 4]
```diff
@@ -52,11 +43,22 @@ def invite_create
...
+    # Get companies that are owned by the logged in user.
+    @companies = Company.where(owner_id: @user.id)
+
+    # Get Roles
+    @roles     = Role.all
+  end
+
   def invite_create_params
     params.require(:user).permit(:first_name, :last_name, :company_id, :role_id, :designation, :email)
   end
```
Comment (MidhatAsif) [inline app/controllers/users_custom_controller.rb]: 1. Final new line missing at the end.

2. Metrics/Line Length: Length is too long. (More that 80 characters)
Refer @ line number 24, 26, 32, 33, 37 & 56.

File: app/controllers/users_custom_controller.rb [hunk 5]
```diff
@@ -1,32 +1,23 @@
...
     new_user_params = invite_create_params
 
     # Create a new user.
     @new_user = User.new(new_user_params)
     @new_user.is_invited_user = true # Mark this user as invited.
-    @new_user.password = generate_password
+    @new_user.password = generate_random_password
 
     # Check if company id is present.
     if new_user_params[:company_id].present?
```
Comment (MidhatAsif) [inline app/controllers/users_custom_controller.rb]: Kindly review the condition: `if current_user.company_id_valid?(new_user_params[:company_id])`. It seems it should be reversed

Comment (MidhatAsif) [inline app/controllers/users_custom_controller.rb]: Can we minimize the branches in this function? (Assignment branch condition size for invite_create is too high)

**Comment by MidhatAsif** [comment]
Review: Cycle-1

File: Gemfile [hunk 6]
```diff
@@ -46,3 +46,6 @@ group :development, :test do
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'end
```
[the remark this answers is not in the document]
Comment (abdullah-siddiqui-git) [reply Gemfile]: Done.

File: app/controllers/users/registrations_controller.rb [hunk 7]
```diff
@@ -1,65 +1,16 @@
...
-  # def after_sign_up_path_for(resource)
-  #   super(resource)
-  # end
-
-  # The path used after sign up for inactive accounts.
-  # def after_inactive_sign_up_path_for(resource)
-  #   super(resource)
-  # end
-
   private
```
[the remark this answers is not in the document]
Comment (abdullah-siddiqui-git) [reply app/controllers/users/registrations_controller.rb]: Noted.

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (abdullah-siddiqui-git) [reply app/views/layouts/application.html.erb]: Done.

File: app/controllers/users_custom_controller.rb (hunk 4 above)
Comment (abdullah-siddiqui-git) [reply app/controllers/users_custom_controller.rb]: Noted.

File: app/controllers/users_custom_controller.rb (hunk 5 above)
Comment (abdullah-siddiqui-git) [reply app/controllers/users_custom_controller.rb]: Incorporated changes.

Comment (abdullah-siddiqui-git) [reply app/controllers/users_custom_controller.rb]: I think its fine. Every assignment in this function seems mandatory.

File: Gemfile [hunk 8]
```diff
@@ -46,3 +46,6 @@ group :development, :test do
...
 
+  gem 'pry'
```
Comment (Anmol79) [inline Gemfile]: fix end tag alignment, also the pry rails should be in development group.

File: app/assets/stylesheets/users_custom.scss [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the users_custom controller here.
```
Comment (Anmol79) [inline app/assets/stylesheets/users_custom.scss]: remove unused files.

File: app/controllers/users/registrations_controller.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+class Users::RegistrationsController < Devise::RegistrationsController
+  # GET /resource/sign_up
+  def new
+    @id_of_administrator_role = Role.where(name: 'Administrator').first.id
```
Comment (Anmol79) [inline app/controllers/users/registrations_controller.rb]: have method in role's model to return the administrator role and should use find as only one admin role will be there

File: test/controllers/users_custom_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+require 'test_helper'
```
Comment (Anmol79) [inline test/controllers/users_custom_controller_test.rb]: remove

File: app/controllers/users/confirmations_controller.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+class Users::ConfirmationsController < Devise::ConfirmationsController
```
Comment (Anmol79) [inline app/controllers/users/confirmations_controller.rb]: remove

File: app/controllers/users/registrations_controller.rb (hunk 10 above)
Comment (Anmol79) [inline app/controllers/users/registrations_controller.rb]: you don't need to get it from the view. just assign the user admin role in the controller.

File: app/controllers/users_custom_controller.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  private
+
+  def set_invitation_view_variables
+    # Get logged in User
+    @user      = current_user
+
+    # Get companies that are owned by the logged in user.
+    @companies = Company.where(owner_id: @user.id)
```
Comment (Anmol79) [inline app/controllers/users_custom_controller.rb]: Don't get all the companies. Should be resolved after current_tenant functionality.

File: app/controllers/users_custom_controller.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    # Create a new user.
+    @new_user = User.new(new_user_params)
+    @new_user.is_invited_user = true # Mark this user as invited.
+    @new_user.password = generate_random_password
+
+    # Check if company id is present.
+    if new_user_params[:company_id].present?
+      # Make sure that provided company id belongs to the set of companies that are owned by logged in user.
+      if current_user.company_id_valid?(new_user_params[:company_id])
+        redirect_to new_user_invite_path, flash: { failure_notification: "Error 403 Forbidden. You tried to access Company you don't own." }
```
Comment (Anmol79) [inline app/controllers/users_custom_controller.rb]: Have flash error, notice and alert. Keep this consistent throughout the project. Error should be an array.
and set the flash like this flash[:error] = ''

File: app/models/user.rb [hunk 15]
```diff
@@ -22,7 +26,26 @@ class User < ActiveRecord::Base
...
-  devise :database_authenticatable, :registerable,
+  # :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable, :confirmable,
          :recoverable, :rememberable, :trackable, :validatable
+
+  def send_invitation_email(company, role)
+    UserMailer.invite(company, self, role).deliver
+  end
+
+  def company_id_valid?(company_id_provided)
```
Comment (Anmol79) [inline app/models/user.rb]: just set the current company, remove this method

File: app/models/user.rb [hunk 16]
```diff
@@ -22,7 +26,26 @@ class User < ActiveRecord::Base
...
+    # Check if index of company_id_provided is nil or not.
+    owned_companies_set = Company.select(:id).where(owner_id: id).collect
+    if owned_companies_set.include? company_id_provided
```
Comment (Anmol79) [inline app/models/user.rb]:
```
if owned_companies_set.include? company_id_provided
  true
else
  false
end
``` ->
```
owned_companies_set.include? company_id_provided
```

File: app/views/devise/shared/_display_devise_login_errors.html.erb [hunk 17]
```diff
@@ -0,0 +1,10 @@
+<!-- DEVISE ONLY PUTS DATA INTO FLASH DICT. USING KEYS 'notice' OR 'alert' -->
+<% flash.each do |key, value| %>
+  <% if key == 'notice' || key == 'alert' %>
+    <div class="alert alert-danger">
+      <a href="#" class="close" data-dismiss="alert">&times;</a>
```
Comment (Anmol79) [inline app/views/devise/shared/_display_devise_login_errors.html.erb]: # -> #_

File: app/views/users_custom/invite.html.erb [hunk 18]
```diff
@@ -0,0 +1,42 @@
+<div class="row">
+  <div class="offset-sm-2 col-sm-8">
+    <%= render 'shared/success_notification' %>
+    <%= render 'shared/failure_notification' %>
+
+    <h2>Invite a User</h2>
+    <%= form_for(@new_user, url: new_user_invite_create_path, method: :post) do |f| %>
+
+      <!-- DISPLAY ERRORS MESSAGES ON FORM SUBMISSION -->
+      <%= render 'shared/errors', obj: @new_user %>
```
Comment (Anmol79) [inline app/views/users_custom/invite.html.erb]: pass the errors array not the object, like errors: @new_user.errors

File: app/views/users_custom/invite.html.erb [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+      </div>
+
+      <div class="field form-group">
+        <%= f.label :email %><br />
+        <%= f.email_field :email, autofocus: true, class: "form-control form-control-sm", placeholder: "john.doe@gmail.com" %>
+      </div>
+
+      <div class="form-group">
+        <label for="role_select">Role:</label>
+        <%= f.select(:role_id, @roles.collect {|role| [role.name, role.id] }, { include_blank: 'None' }, class: 'form-control form-control-sm', id: 'role_select') %>
```
Comment (Anmol79) [inline app/views/users_custom/invite.html.erb]: space and create a helper for this ```@roles.collect { |role| [role.name, role.id] }```

File: app/assets/stylesheets/users_custom.scss (hunk 9 above)
Comment (abdullah-siddiqui-git) [reply app/assets/stylesheets/users_custom.scss]: Resolved in next PR.

File: app/controllers/users/confirmations_controller.rb (hunk 12 above)
Comment (abdullah-siddiqui-git) [reply app/controllers/users/confirmations_controller.rb]: Cannot remove this file as Devise, under the hood, uses it to refer to the methods present in the parent class.

File: app/controllers/users/registrations_controller.rb (hunk 10 above)
Comment (abdullah-siddiqui-git) [reply app/controllers/users/registrations_controller.rb]: - Resolved by creating a static method in Role model which returns object of "Role" whenever called like:
admin = Role.admin # admin is an object related to Role model. 
- Role ID fetching from view mechanism has been eliminated. While creation of new User, admin role id is fetched and set as role_id of new User.
- Upon calling super before setting a variable causes an error as super tries to render the view before the form object is even created.

Comment (abdullah-siddiqui-git) [reply app/controllers/users/registrations_controller.rb]: Resolved.

File: app/models/user.rb (hunk 15 above)
Comment (abdullah-siddiqui-git) [reply app/models/user.rb]: Resolved in the next PR.

File: app/views/users_custom/invite.html.erb (hunk 19 above)
Comment (abdullah-siddiqui-git) [reply app/views/users_custom/invite.html.erb]: Resolved in the next PR.

File: test/controllers/users_custom_controller_test.rb (hunk 11 above)
Comment (abdullah-siddiqui-git) [reply test/controllers/users_custom_controller_test.rb]: Done in the next PR.

File: app/models/user.rb (hunk 16 above)
Comment (abdullah-siddiqui-git) [reply app/models/user.rb]: Resolved in the next PR.

File: app/views/users_custom/invite.html.erb (hunk 18 above)
Comment (abdullah-siddiqui-git) [reply app/views/users_custom/invite.html.erb]: Resolved in the next PR.

File: app/views/devise/shared/_display_devise_login_errors.html.erb (hunk 17 above)
Comment (abdullah-siddiqui-git) [reply app/views/devise/shared/_display_devise_login_errors.html.erb]: Resolved in the next PR.

**Pull request merged**